CASSANDRA-14405 Transient Replication: Metadata refactor

bdeggleston opened the pull request.

bdeggleston committed:
adding Replica and ReplicatedRange classes

bdeggleston committed:
fixing replica equality

bdeggleston committed:
adding transient replication enable to config

bdeggleston committed:
adding RF tests

bdeggleston committed:
updating rf

bdeggleston committed:
adding initial nts support

bdeggleston committed:
renaming endpoint specific names

bdeggleston committed:
adding simple strategy support

bdeggleston committed:
prevent mixing of transient replication and MV/2i

bdeggleston committed:
fixing some tests

bdeggleston committed:
more test fixing

bdeggleston committed:
fixing hint dtest

bdeggleston committed:
fixing rf comparison

bdeggleston committed:
adding range to replica

bdeggleston committed:
fixing move test

bdeggleston committed:
merging Replica and ReplicatedRange

bdeggleston committed:
fixing range tests

bdeggleston committed:
more test fixing

bdeggleston committed:
fixing srp test, misc adjustments

bdeggleston committed:
fixing concurrent list modification

bdeggleston committed:
fixing MV test

bdeggleston committed:
renaming replicas to replica helpers

bdeggleston committed:
adding replica aware collections

bdeggleston committed:
moving stuff out of ReplicaHelpers

bdeggleston committed:
finishing move away from replica helpers

bdeggleston committed:
undoing multimap changes

bdeggleston committed:
fixing tests

bdeggleston committed:
fixing decommission

bdeggleston committed:
adding config and docs

aweisberg commented on doc/source/architecture/dynamo.rst:
```diff
@@ -74,6 +74,26 @@ nodes in each rack, the data load on the smallest rack may be much higher.  Simi
+
+Transient replication allows you to configure a subset of replicas to only replicate data that hasn't been incrementally
+repaired. This allows you to trade data redundancy for storage usage, and increased read and write throughput. For instance,
+if you have a replication factor of 3, with 1 transient replica, 2 replicas will replicate all data for a given token
```
Use the 3 replicas upgraded to 5 case. It's not a tradeoff of redundancy at all. We can't message that at all and it's not true.

bdeggleston replied:
fixed

aweisberg commented on doc/source/architecture/dynamo.rst:
```diff
@@ -74,6 +74,26 @@ nodes in each rack, the data load on the smallest rack may be much higher.  Simi
+repaired. This allows you to trade data redundancy for storage usage, and increased read and write throughput. For instance,
+if you have a replication factor of 3, with 1 transient replica, 2 replicas will replicate all data for a given token
+range, while the 3rd will only keep data that hasn't been incrementally repaired. Since you're reducing the copies kept
+of data by the number of transient replicas, transient replication is best suited to multiple dc deployments.
```
No you don't need multiple DCs!

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/cql3/statements/AlterKeyspaceStatement.java:
```diff
@@ -96,7 +98,35 @@ private void warnIfIncreasingRF(KeyspaceMetadata ksm, KeyspaceParams params)
+            }
+
+        }
+    }
```
I think we can allow people to increase # of transient replicas (they make it real with nodetool cleanup or repair), but how does decreasing number of transient replicas work safely? I think we should disallow it until we can articulate how and why it is safe. There has to be a temporary pending state where a node is transitioning from transient to full where it receives writes but not reads because it can't correctly service reads as a full replcia.

bdeggleston replied:
If you're keeping the total number of replicas the same, I don't think there would  be any difference between reducing the number of transient replicas, and increasing the replication factor on a normally replicated keyspace. You'd just decrease transient count one at a time and do a full repair between each? If you were reducing both (ie: 5/2 -> 4/1), that shouldn't cause any problems at all.

aweisberg replied:
I've never actually thought about the procedure for changing RF. I think I get the principle behind one at a time changes.

What happens when you write at CL.ALL and read at CL.ONE? What happens for ranges where the node that is now a full replica but doesn't have the full data is read from?

aweisberg replied:
Sorry I was reading your example wrong. I didn't realize the number of full replicas was staying the same.

I don't think a full repair is necessary?

bdeggleston replied:
> I don't think a full repair is necessary?

If you don't, it won't get he repaired data it needs.

aweisberg replied:
If you are decreasing the transient count who the full replicas are is unchanged? So you don't need to move data around? You just need to fetch what you are about to lose from the transients before you lose another replica?

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicationFactor.java:
```diff
@@ -0,0 +1,121 @@
+
+    public final int trans;
+    public final int replicas;
+    public transient final int full;
```
Why is this field transient?

bdeggleston replied:
I think I did that as sort of a hint that it's always calculated in the ctor as a convenience. Since rf is defined as `all_replicas/trans_replicas`, and the `full` value is never provided by the caller.

aweisberg commented on src/java/org/apache/cassandra/cql3/statements/AlterKeyspaceStatement.java:
```diff
@@ -96,7 +98,35 @@ private void warnIfIncreasingRF(KeyspaceMetadata ksm, KeyspaceParams params)
+
+    private void warnIfIncreasingRF(AbstractReplicationStrategy oldStrategy, AbstractReplicationStrategy newStrategy)
+    {
+        if (newStrategy.getReplicationFactor().full > oldStrategy.getReplicationFactor().full)
```
Should this be the number of full replicas or the total number of replicas?

bdeggleston replied:
I think full is what we want here. Increasing the number of full replicas would require a repair. If the total rf is increased, but the number of full replicas doesn't change, then you're just adding transient replicas, which shouldn't require a repair or any other operator action.

aweisberg replied:
Huh, so because your quorum size increases you don't need to run even an incremental repair? Is that always true. Have to think about it some more.

But... while it might be safe to do one increase, if you increased transient replication twice in sequence and it didn't change th quorum size could you get into trouble?

bdeggleston replied:
Since replication factor changes don't really fall under any of the existing JIRAs, maybe we should disable all rf changes (except transient increases) as part of the initial refactor, and make a follow on jira to enable the other rf changes after the read and write path changes are in. Then we'll be able to reason about them in isolation, and will be better equipped to do the needed testing / tool updates.

aweisberg replied:
Sounds good.

aweisberg commented on src/java/org/apache/cassandra/db/ColumnFamilyStore.java:
```diff
@@ -1591,7 +1589,7 @@ public long getExpectedCompactedFileSize(Iterable<SSTableReader> sstables, Opera
         // cleanup size estimation only counts bytes for keys local to this node
         long expectedFileSize = 0;
-        Collection<Range<Token>> ranges = StorageService.instance.getLocalRanges(keyspace.getName());
+        Collection<Range<Token>> ranges = StorageService.instance.getLocalReplicas(keyspace.getName()).asRangeSet();
```
Since this is a collection could you use Collections2.transform instead of materializing a set?

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/batchlog/BatchlogManager.java:
```diff
@@ -490,16 +497,16 @@ private static int gcgs(Collection<Mutation> mutations)
-                super(writeEndpoints, Collections.<InetAddressAndPort>emptySet(), null, null, null, WriteType.UNLOGGED_BATCH, queryStartNanoTime);
-                undelivered.addAll(writeEndpoints);
+                super(writeReplicas, ReplicaList.of(), null, null, null, WriteType.UNLOGGED_BATCH, queryStartNanoTime);
+                Iterables.addAll(undelivered, writeReplicas.asEndpoints());
```
This is a case where it seems like we could call a Collections2.transform version and not have to create a new list. One thing to keep in mind is that Collections2.transform makes contains and remove slow because it's O(n) no matter what the underlying collection is.

bdeggleston replied:
asEndpoints / asRanges doesn't allocate a new list, it just transforms the iterable

aweisberg commented on src/java/org/apache/cassandra/db/ColumnFamilyStore.java:
```diff
@@ -1868,7 +1866,7 @@ public void compactionDiskSpaceCheck(boolean enable)
     public void cleanupCache()
     {
-        Collection<Range<Token>> ranges = StorageService.instance.getLocalRanges(keyspace.getName());
+        Collection<Range<Token>> ranges = StorageService.instance.getLocalReplicas(keyspace.getName()).asRangeSet();
```
Same.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/db/ConsistencyLevel.java:
```diff
@@ -190,50 +197,50 @@ public int countLocalEndpoints(Iterable<InetAddressAndPort> liveEndpoints)
+        Map<String, ReplicaList> dcsReplicas = new HashMap<>();
         for (String dc: strategy.getDatacenters())
-            dcsEndpoints.put(dc, new ArrayList<>());
+            dcsReplicas.put(dc, new ReplicaList());
```
Completely overkill and unrelated, but ArrayList will over allocate here if liveReplicas is < the size of the default allocation of 10

bdeggleston replied:
why not? fixed

aweisberg commented on src/java/org/apache/cassandra/db/ConsistencyLevel.java:
```diff
@@ -190,50 +197,50 @@ public int countLocalEndpoints(Iterable<InetAddressAndPort> liveEndpoints)
 
-        List<InetAddressAndPort> waitSet = new ArrayList<>();
-        for (Map.Entry<String, List<InetAddressAndPort>> dcEndpoints : dcsEndpoints.entrySet())
+        ReplicaList waitSet = new ReplicaList();
```
Another potential over allocation.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/service/reads/DataResolver.java:
```diff
@@ -64,12 +72,19 @@ public PartitionIterator resolve()
-            sources[i] = msg.from;
+
+            Replica replica = replicaMap.get(msg.from);
+            if (replica == null)
```
It seems like we knew what kind of Replica it was when we sent the message. The kind of Replica it might be when we get the response could be different? Probably not given how pending states work.

I just don't get how we could fail to decorate it. That's not right. If we sent the message eliciting this response then we must have known at the time.

bdeggleston replied:
My thinking here was that the speculative read repair patch (in review still) could speculate to a node right before a ring change, and then wouldn't be able to find it through the read command. Since the speculation and read responses will be on different threads, adding replicas post DataResolver instantiation would require synchronization of the map, which isn't really worth it for the vanishingly small chance of this race occurring. The consequence of using a fake full replica is just the potential sending of read repair mutations to a transient replica.

aweisberg commented on src/java/org/apache/cassandra/service/reads/DataResolver.java:
```diff
@@ -64,12 +72,19 @@ public PartitionIterator resolve()
+            if (replica == null)
+                replica = Replica.fullStandin(msg.from);
+
+            sources[i] = replica != null ? replica : Replica.fullStandin(msg.from);
```
Seems like you doubled up on the stand in?

aweisberg commented on src/java/org/apache/cassandra/locator/Replica.java:
```diff
@@ -0,0 +1,221 @@
+     * We need to assume an endpoint is a full replica in a with unknown ranges in a
+     * few cases, so this returns one that throw an exception if you try to get it's range
+     */
+    public static Replica fullStandin(InetAddressAndPort endpoint)
```
This feels so unsafe. At the first level it's turning a compile time error into a runtime error. But then it's turning the fullness or not of the replica into a silent error. Do things work if you have isFull and isTransient throw as well?

Another issue is this is going to make isFull() and isTransient() bi-morphic which isn't the best. Feels like it might be premature optimization to worry about that right now though.

bdeggleston replied:
I agree it sucks. The problem is that there batchlog and hint related functions that don't work particularly well with the Replica idea, but use the snitch and some write path stuff that were converted to use Replicas to prevent a lot of conversion/copying on the hot path. So using fake replicas solved that problem, but not that well. Let me take another look at the places that use stand ins though, there's probably something less crappy that can be done.

aweisberg commented on src/java/org/apache/cassandra/db/ReadCommand.java:
```diff
@@ -128,6 +130,7 @@ protected ReadCommand(Kind kind,
 
     protected abstract void serializeSelection(DataOutputPlus out, int version) throws IOException;
     protected abstract long selectionSerializedSize(int version);
+    public abstract Replica decorateEndpoint(InetAddressAndPort endpoint);
```
I'm not sure if this is a good idea based on it being used after we have already committed to that endpoint replicating a specific range a specific way. This seems like it races with ring changes.

aweisberg commented on src/java/org/apache/cassandra/db/compaction/CompactionManager.java:
```diff
@@ -533,7 +530,7 @@ public AllSSTableOpStatus relocateSSTables(final ColumnFamilyStore cfs, int jobs
             return AllSSTableOpStatus.ABORTED;
         }
-        final Collection<Range<Token>> r = StorageService.instance.getLocalRanges(cfs.keyspace.getName());
+        final Collection<Range<Token>> r = StorageService.instance.getLocalReplicas(cfs.keyspace.getName()).asRangeSet();
```
This is hardly even used. It just checks for isEmpty(). No need to unwrap.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/db/compaction/CompactionManager.java:
```diff
@@ -468,7 +465,7 @@ public AllSSTableOpStatus performCleanup(final ColumnFamilyStore cfStore, int jo
         }
         // if local ranges is empty, it means no data should remain
-        final Collection<Range<Token>> ranges = StorageService.instance.getLocalRanges(keyspace.getName());
+        final Collection<Range<Token>> ranges = StorageService.instance.getLocalReplicas(keyspace.getName()).asRangeSet();
```
This could use a Collection2.transform version. It's just iterated.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/db/compaction/CompactionManager.java:
```diff
@@ -871,7 +868,7 @@ public void forceUserDefinedCleanup(String dataFiles)
             ColumnFamilyStore cfs = entry.getKey();
             Keyspace keyspace = cfs.keyspace;
-            Collection<Range<Token>> ranges = StorageService.instance.getLocalRanges(keyspace.getName());
+            Collection<Range<Token>> ranges = StorageService.instance.getLocalReplicas(keyspace.getName()).asRangeSet();
```
Also just iterated

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/db/compaction/Verifier.java:
```diff
@@ -209,7 +208,9 @@ public void verify()
-            List<Range<Token>> ownedRanges = isOffline ? Collections.emptyList() : Range.normalize(StorageService.instance.getLocalAndPendingRanges(cfs.metadata().keyspace));
+            List<Range<Token>> ownedRanges = isOffline
+                                             ? Collections.emptyList()
+                                             : Range.normalize(StorageService.instance.getLocalAndPendingReplicas(cfs.metadata().keyspace).asRangeSet());
```
Another candidate

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/db/view/ViewBuilder.java:
```diff
@@ -135,14 +137,15 @@ private synchronized void build()
-
+        ReplicaSet replicatedRanges = StorageService.instance.getLocalReplicas(ksName);
+        Replicas.checkFull(StorageService.instance.getLocalReplicas(ksName));
+        Set<Range<Token>> newRanges = replicatedRanges.asRangeSet()
```
Also just iterating here. You don't even need Collection2 you can unwrap it in the first map.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/dht/RangeFetchMapCalculator.java:
```diff
@@ -158,14 +159,15 @@ static boolean isTrivial(Range<Token> range)
+                    if (passFilters(replica, localDCCheck))
                     {
-                        fetchMap.put(src, trivialRange);
+                        fetchMap.put(replica.getEndpoint(), trivialRange);
```
So this is something I have been running into on my end, but why unwrap here?  I have been need to fix that because I need the transientness of what I am fetching for to determine whether the remote side is going to send me the transient data or the full data.

aweisberg commented on src/java/org/apache/cassandra/dht/RangeFetchMapCalculator.java:
```diff
@@ -347,15 +349,16 @@ else if (vertex.isRangeVertex())
                 sourceFound = true;
                 // if we pass filters, it means that we don't filter away localhost and we can count it as a source:
-                if (endpoint.equals(FBUtilities.getBroadcastAddressAndPort()))
+                if (replica.getEndpoint().equals(FBUtilities.getBroadcastAddressAndPort()))
```
Crazy thought, what if Replica had an isLocal or something?

bdeggleston replied:
good idea, changed

aweisberg commented on src/java/org/apache/cassandra/locator/Replicas.java:
```diff
@@ -0,0 +1,313 @@
+        return result;
+    }
+
+    public Iterable<Range<Token>> fullRanges()
```
Is this really a good thing to have around? The one use I see of this doesn't look appropriate.

aweisberg commented on src/java/org/apache/cassandra/dht/RangeStreamer.java:
```diff
@@ -176,25 +179,28 @@ public void addSourceFilter(ISourceFilter filter)
+     * @param replicas ranges to be streamed
      */
-    public void addRanges(String keyspaceName, Collection<Range<Token>> ranges)
+    public void addRanges(String keyspaceName, Replicas replicas)
```
Should this be addReplicas? I would also change the comment to ranges to be fetched. Streaming seems to get used to imply both sending and fetching. But fetching always means fetching.

aweisberg commented on src/java/org/apache/cassandra/dht/RangeStreamer.java:
```diff
@@ -176,25 +179,28 @@ public void addSourceFilter(ISourceFilter filter)
-        Multimap<Range<Token>, InetAddressAndPort> rangesForKeyspace = useStrictSource
-                ? getAllRangesWithStrictSourcesFor(keyspaceName, ranges) : getAllRangesWithSourcesFor(keyspaceName, ranges);
+        ReplicaMultimap<Range<Token>, ReplicaList> rangesForKeyspace = useStrictSource
+                                                                       ? getAllRangesWithStrictSourcesFor(keyspaceName, replicas.fullRanges())
```
fullRanges(), not a fan.

aweisberg commented on src/java/org/apache/cassandra/locator/AbstractEndpointSnitch.java:
```diff
@@ -23,17 +23,17 @@
 public abstract class AbstractEndpointSnitch implements IEndpointSnitch
 {
-    public abstract int compareEndpoints(InetAddressAndPort target, InetAddressAndPort a1, InetAddressAndPort a2);
+    public abstract int compareEndpoints(InetAddressAndPort target, Replica r1, Replica r2);
```
Are they endpoints or replicas now? I think maybe this should just be endpoints to make explicit the snitch isn't going to use the range or transient information. It doesn't have many usages so it's not a big deal to manually

jeffjirsa commented on conf/cassandra.yaml:
```diff
@@ -1032,6 +1032,10 @@ enable_scripted_user_defined_functions: false
 enable_materialized_views: true
 
+# Enables creation of transiently replicated keyspaces on this node.
+# Transient replication is experimental and is not recommended for production use.
```
Thanks for the experimental line.

aweisberg commented on src/java/org/apache/cassandra/locator/AbstractReplicationStrategy.java:
```diff
@@ -202,61 +204,63 @@ private Keyspace getKeyspace()
+            for (Replica replica : calculateNaturalReplicas(token, metadata))
             {
-                map.put(ep, range);
+                Preconditions.checkState(range.equals(replica.getRange()) || this instanceof LocalStrategy);
```
Can you explain this check? Why would LocalStrategy not have the ranges match?

bdeggleston replied:
added comment

aweisberg commented on src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java:
```diff
@@ -424,17 +419,17 @@ public boolean isWorthMergingForRangeQuery(List<InetAddressAndPort> merged, List
-        double maxMerged = maxScore(merged);
-        double maxL1 = maxScore(l1);
-        double maxL2 = maxScore(l2);
+        double maxMerged = maxScore(merged.asEndpoints());
```
Maybe just update maxScore?

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaSet.java:
```diff
@@ -0,0 +1,142 @@
+        return new ReplicaSet(ImmutableSet.<Replica>builder().addAll(from).build());
+    }
+
+    public static ReplicaSet ordered()
```
I had to look this up to realize that ordered doesn't mean sorted. It's the correct name, they are ordered just not according to their natural order. Maybe the right shade for the bike shed here is order preserving?

aweisberg commented on src/java/org/apache/cassandra/locator/PendingRangeMaps.java:
```diff
@@ -23,196 +23,176 @@
-                return res;
-            return o1.right.compareTo(o2.right);
-        }
+    private static final Comparator<Range<Token>> descendingComparatorForWrapAround = (o1, o2) -> {
```
These are kind of orthogonal changes, but sure. I don't see a down side to doing them here. I'll bet these comparators are low traffic so it won't be much merge pain.

bdeggleston replied:
:+1:

aweisberg commented on src/java/org/apache/cassandra/locator/PendingRangeMaps.java:
```diff
@@ -23,196 +23,176 @@
+            for (Replica replica : entry.getValue())
             {
-                sb.append(address).append(':').append(range);
+                sb.append(replica).append(':').append(range);
```
What kind of compatibility issues are we creating here by changing the output? (And what kind did I create with InetAddressAndPort?)

Doesn't look like an issue to me since it's used in a trace log statement and TokenMetadata.toString().

bdeggleston replied:
Some, I'm sure. Not sure what we can really do about it though, since we're adding the port and transient data it kind of needs to be included here. At least it's in a major release though?

aweisberg commented on src/java/org/apache/cassandra/locator/Replica.java:
```diff
@@ -0,0 +1,221 @@
+        this(endpoint, new Range<>(start, end), full);
+    }
+
+    public boolean equals(Object o)
```
Was this how an auto-generated implementation? Just want to make sure it's "correct" since it's always tricky to do equals with best practices.

bdeggleston replied:
yep, auto generated

aweisberg commented on src/java/org/apache/cassandra/locator/Replica.java:
```diff
@@ -0,0 +1,221 @@
+        };
+    }
+
+    public static Replica full(InetAddressAndPort endpoint, Token start, Token end)
```
So generally people shouldn't be constructing using full or transient and I think maybe the javadoc should say that.

You almost always want a real source of the transient status not a fake one. Propagating fake transient status is going to get us into trouble.

aweisberg commented on src/java/org/apache/cassandra/locator/Replica.java:
```diff
@@ -0,0 +1,221 @@
+        return trans(endpoint, new Range<>(start, end));
+    }
+
+    public static void assureSufficientFullReplica(Collection<Replica> replicas, ConsistencyLevel cl) throws UnavailableException
```
Unused

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/locator/Replica.java:
```diff
@@ -0,0 +1,221 @@
+        return Iterables.transform(replicas, Replica::getEndpoint);
+    }
+
+    public static List<InetAddressAndPort> toEndpointList(List<Replica> replicas)
```
Unused

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/locator/Replica.java:
```diff
@@ -0,0 +1,221 @@
+        }
+    }
+
+    public static Iterable<InetAddressAndPort> toEndpoints(Iterable<Replica> replicas)
```
I think all of these collection oriented things should move into Replicas. I also think we should make explicit the lazy vs eager transformation options and generally offer both so that code can try and pick the right one.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/locator/Replicas.java:
```diff
@@ -0,0 +1,313 @@
+ * transient replication status, basic contains and remove methods can be ambiguous. Replicas forces you
+ * to be explicit about what you're checking the container for, or removing from it.
+ */
+public abstract class Replicas implements Iterable<Replica>
```
This is really ReplicaCollection. The XYZs idiom is usually an additional class that has additional utility methods.

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaList.java:
```diff
@@ -0,0 +1,244 @@
+    public int hashCode()
+    {
+
+        return Objects.hash(replicaList);
```
replicaList is not null so I would have just used replicaList.hashCode(). This might allocate a wrapper array for varargs.

Objects.hashCode() is the single arg version that won't allocate.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaList.java:
```diff
@@ -0,0 +1,244 @@
+        //   so will be very small (< RF). In that case, retainAll is in fact more efficient.
+        //   2) we do ultimately need a list so converting everything to sets don't make sense
+        //   3) l1 and l2 are sorted by proximity. The use of retainAll  maintain that sorting in the result, while using sets wouldn't.
+        Collection<InetAddressAndPort> endpoints = l2.asEndpointList();
```
You can do this without converting or wrapping because l2 implements containsEndpoint.

It might make sense to implement containsEndpoint here in ReplicaList so that it doesn't have to allocate an iterator. You can't do a similar thing with ReplicaSet since you can't access the Set implementations underlying storage for iteration.

The disadvantage is it makes containsEndpoint bi-morphic which is probably fine.

Granted the iterator is probably allocated on the stack since escape analysis should work in this instance and its methods get inlined so??? 

Also Iterables.any requires allocating the predicate since it's binding a value. Probably worth avoiding that.

aweisberg commented on src/java/org/apache/cassandra/locator/Replica.java:
```diff
@@ -0,0 +1,221 @@
+    {
+        if (isFull() && Iterables.all(toSubtract, Replica::isFull))
+        {
+            Set<Range<Token>> subtractedRanges = getRange().subtractAll(toSubtract.asRangeSet());
```
Range.subtractAll could take an iterable here and then we could use a lazy transformation.

aweisberg commented on src/java/org/apache/cassandra/locator/Replica.java:
```diff
@@ -0,0 +1,221 @@
+
+    public ReplicaList normalizeByRange()
+    {
+        List<Range<Token>> normalized = Range.normalize(Collections.singleton(getRange()));
```
This is a little weird as a way to normalize a single thing? Isn't this basically unwrap and most of the time won't it not unwrap?

Maybe we can return Replicas.singleton most of the time?

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaList.java:
```diff
@@ -0,0 +1,244 @@
+    @Override
+    public void removeEndpoint(InetAddressAndPort endpoint)
+    {
+        replicaList.removeIf(r -> r.getEndpoint().equals(endpoint));
```
This allocates a lambda. Better to iterate and remove by index. removeIf also allocates an iterator. Also you need to handle a null endpoint correctly. Does ReplicaList and Replicas allow null entries?

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaList.java:
```diff
@@ -0,0 +1,244 @@
+    public ReplicaList filter(Predicate<Replica> predicate)
+    {
+        ArrayList<Replica> newReplicaList = new ArrayList<>(size());
+        for (Replica replica: replicaList)
```
Iterate by index rather than using an iterator.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaList.java:
```diff
@@ -0,0 +1,244 @@
+
+    public ReplicaList filter(Predicate<Replica> predicate)
+    {
+        ArrayList<Replica> newReplicaList = new ArrayList<>(size());
```
This only makes sense if size() is smaller than the default allocation size of ArrayList? Otherwise you don't know what will come out of filter and probably want to use the usual algorithm?

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaList.java:
```diff
@@ -0,0 +1,244 @@
+    {
+        if (Iterables.all(this, Replica::isFull))
+        {
+            ReplicaList normalized = new ReplicaList(size());
```
Maybe you want to invert things a bit and pass this into each call to replica.normalizeByRange?

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaMultimap.java:
```diff
@@ -0,0 +1,121 @@
+
+import com.google.common.collect.Iterables;
+
+public abstract class ReplicaMultimap<K, V extends Replicas>
```
Doesn't implement equals or hashCode? That's going to surprise someone eventually.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaSet.java:
```diff
@@ -0,0 +1,142 @@
+
+    public int hashCode()
+    {
+        return Objects.hash(replicaSet);
```
Uses the varargs version

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaSet.java:
```diff
@@ -0,0 +1,142 @@
+    @Override
+    public void removeEndpoint(InetAddressAndPort endpoint)
+    {
+        replicaSet.removeIf(r -> r.getEndpoint().equals(endpoint));
```
Allocates a lambda.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaList.java:
```diff
@@ -0,0 +1,244 @@
+        return replicaList.get(idx);
+    }
+
+    public List<InetAddressAndPort> asEndpointList()
```
This is in Replicas which I think is the better place since it allows conversion from any Replicas to an endpoint list.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/locator/Replicas.java:
```diff
@@ -0,0 +1,313 @@
+        return Iterables.transform(this, Replica::getEndpoint);
+    }
+
+    public Set<InetAddressAndPort> asEndpointSet()
```
I am thinking we should have two versions, asEndpointSet/asRangeSet and toEndpointSet/toRangeSet. The "to" version allocates and converts. The "as" version is a Collections2 wrapper.

I think in a lot of places we will want neither because I saw several opportunities to manually unwrap.

aweisberg commented on src/java/org/apache/cassandra/locator/Replicas.java:
```diff
@@ -0,0 +1,313 @@
+
+    public static Replicas singleton(Replica replica)
+    {
+        return of(Collections.singleton(replica));
```
This could be one less allocation if you skip the Collections.singleton.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/locator/Replicas.java:
```diff
@@ -0,0 +1,313 @@
+
+    public boolean containsEndpoint(InetAddressAndPort endpoint)
+    {
+        return Iterables.any(this, r -> r.getEndpoint().equals(endpoint));
```
This allocates a lambda. Also ReplicaList can implement this without allocating an iterator.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/locator/TokenMetadata.java:
```diff
@@ -856,25 +857,25 @@ private static PendingRangeMaps calculatePendingRanges(AbstractReplicationStrate
         Set<Range<Token>> affectedRanges = new HashSet<Range<Token>>();
         for (InetAddressAndPort endpoint : leavingEndpoints)
-            affectedRanges.addAll(addressRanges.get(endpoint));
+            affectedRanges.addAll(addressRanges.get(endpoint).asRangeSet());
```
No reason to allocate a new set here, this is a candidate for Collections2.transform

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/locator/TokenMetadata.java:
```diff
@@ -1204,21 +1205,21 @@ private String printPendingRanges()
         PendingRangeMaps pendingRangeMaps = this.pendingRanges.get(keyspaceName);
         if (pendingRangeMaps == null)
-            return Collections.emptyList();
+            return new ReplicaList(0);
```
Use a singleton immutable empty list?

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/net/MessagingService.java:
```diff
@@ -591,8 +592,9 @@ public void run()
-                    Mutation mutation = ((WriteCallbackInfo) expiredCallbackInfo).mutation();
-                    return StorageProxy.submitHint(mutation, expiredCallbackInfo.target, null);
+                    WriteCallbackInfo writeCallbackInfo = ((WriteCallbackInfo) expiredCallbackInfo);
+                    Mutation mutation = writeCallbackInfo.mutation();
```
Kind of a rando change, were you looking at mutation in a debugger?

bdeggleston replied:
we need to get both the mutation and the replica out of the write callback info here

aweisberg commented on src/java/org/apache/cassandra/service/AbstractWriteResponseHandler.java:
```diff
@@ -225,7 +223,7 @@ protected boolean waitingFor(InetAddressAndPort from)
     public void assureSufficientLiveNodes() throws UnavailableException
     {
-        consistencyLevel.assureSufficientLiveNodes(keyspace, Iterables.filter(Iterables.concat(naturalEndpoints, pendingEndpoints), isAlive));
+        consistencyLevel.assureSufficientLiveNodes(keyspace, Replicas.filter(Replicas.concatNaturalAndPending(naturalReplicas, pendingReplicas), isReplicaAlive));
```
This ends up allocating two containers when all that is wanted is Iterables.

bdeggleston replied:
I don't think we're actually allocating any more objects than we were before. Both `filter` and `concat`/`concatNaturalAndPending` each allocate a single iterable/replicas warpper, and don't copy anything.

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaList.java:
```diff
@@ -0,0 +1,244 @@
+        }
+    }
+
+    public static ReplicaList immutableCopyOf(Replicas replicas)
```
You don't need to allocate a ReplicaList if the replicas is already a replica list containing an ImmutableList. You can just return what was passed to you. This is an optimization Immutabe* do so that it's low overhead to ask for defensive copies.

aweisberg commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -344,47 +343,43 @@ private static void recordCasContention(int contentions)
-        List<InetAddressAndPort> naturalEndpoints = StorageService.instance.getNaturalEndpoints(metadata.keyspace, tk);
-        Collection<InetAddressAndPort> pendingEndpoints = StorageService.instance.getTokenMetadata().pendingEndpointsFor(tk, metadata.keyspace);
+        ReplicaList naturalReplicas = StorageService.instance.getNaturalReplicas(metadata.keyspace, tk);
+        ReplicaList pendingReplicas = new ReplicaList(StorageService.instance.getTokenMetadata().pendingEndpointsFor(tk, metadata.keyspace));
```
Interesting you end up having to make a copy to a list to convert it from Replicas (actually a set if you get a non-empty value back).

aweisberg commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -503,12 +498,12 @@ private static void sendCommit(Commit commit, Iterable<InetAddressAndPort> repli
         PrepareCallback callback = new PrepareCallback(toPrepare.update.partitionKey(), toPrepare.update.metadata(), requiredParticipants, consistencyForPaxos, queryStartNanoTime);
         MessageOut<Commit> message = new MessageOut<Commit>(MessagingService.Verb.PAXOS_PREPARE, toPrepare, Commit.serializer);
-        for (InetAddressAndPort target : endpoints)
+        for (InetAddressAndPort target : replicas.asEndpoints())
```
This is a candidate for manually unwrapping in the loop.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -541,12 +536,12 @@ public void run()
+        ProposeCallback callback = new ProposeCallback(replicas.size(), requiredParticipants, !timeoutIfPartial, consistencyLevel, queryStartNanoTime);
         MessageOut<Commit> message = new MessageOut<Commit>(MessagingService.Verb.PAXOS_PROPOSE, proposal, Commit.serializer);
-        for (InetAddressAndPort target : endpoints)
+        for (InetAddressAndPort target : replicas.asEndpoints())
```
Another candidate for manually unwrapping.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaList.java:
```diff
@@ -0,0 +1,244 @@
+        return l1.filter(r -> endpoints.contains(r.getEndpoint()));
+    }
+
+    public static ReplicaList of(Replica... replicas)
```
Some good specializations of this would be 0, 1 and several like in ImmutableList

bdeggleston replied:
fixed

aweisberg replied:
Hmmm so what I thought was a specialization would be possible, but since it's mutable you have to allocate a real list. So really I think it's fine to have of() allocate the wrapper array. So of() could just return a global immutable one. The singleton one could do a singleton list wrapper.

TL;DR this is not important enough to use up more of your time. You can delete the specializations.

aweisberg commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -1098,7 +1095,7 @@ private static void asyncRemoveFromBatchlog(Collection<InetAddressAndPort> endpo
 
             if (canDoLocalRequest(target))
-                performLocally(Stage.MUTATION, () -> BatchlogManager.remove(uuid));
+                performLocally(Stage.MUTATION, Replica.fullStandin(target), () -> BatchlogManager.remove(uuid));
```
Full stand in! Either a replica is necessary downstream or it isn't. I checked and to me it sure looks necessary because we use this to decide who to hint to and the decision of should we hint or not is based on the transient state.

aweisberg commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -1364,68 +1363,72 @@ public static void sendToHintedEndpoints(final Mutation mutation,
         }
-        message = message.withParameter(ParameterType.FORWARD_TO.FORWARD_TO, new ForwardToContainer(targets, messageIds));
+        Replicas.checkFull(targets);
+        message = message.withParameter(ParameterType.FORWARD_TO.FORWARD_TO, new ForwardToContainer(targets.asEndpointList(), messageIds));
```
A candidate for lazily converting.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -3863,17 +3863,12 @@ public void forceTerminateAllRepairSessions()
-            if (FailureDetector.instance.isAlive(endpoint))
-                liveEps.add(endpoint);
-        }
+        return replicas.filter(r -> FailureDetector.instance.isAlive(r.getEndpoint()));
```
if isAlive were static you wouldn't have to allocate lambda here.

aweisberg commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -3845,16 +3852,9 @@ public void forceTerminateAllRepairSessions()
-    }
-
-    public List<InetAddressAndPort> getLiveNaturalEndpoints(Keyspace keyspace, RingPosition pos)
+    public ReplicaList getLiveNaturalReplicas(Keyspace keyspace, ByteBuffer key)
```
Unused

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -3863,17 +3863,12 @@ public void forceTerminateAllRepairSessions()
+    public ReplicaList getLiveNaturalReplicas(Keyspace keyspace, RingPosition pos)
     {
-        List<InetAddressAndPort> endpoints = keyspace.getReplicationStrategy().getNaturalEndpoints(pos);
+        ReplicaList replicas = keyspace.getReplicationStrategy().getNaturalReplicas(pos);
```
If getNaturalReplicas took a filter we could filter out the replicas into the copied list that getNaturalReplicas already creates.

aweisberg commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -4307,32 +4300,38 @@ private void calculateToFromStreams(Collection<Token> newTokens, List<String> ke
+
                         logger.debug("Range: {} Current endpoints: {} New endpoints: {}", toStream, currentEndpoints, newEndpoints);
-                        for (InetAddressAndPort address : Sets.difference(newEndpoints, currentEndpoints))
+                        for (Replica replica : Sets.difference(newEndpoints, currentEndpoints))
```
I am not sure this works since it should be by address. I think I found cases that were suspect when it also compared by range. 

I thought we weren't going to do regular collections of Replicas for the most part so it would be explicit?

aweisberg commented on src/java/org/apache/cassandra/service/WriteResponseHandler.java:
```diff
@@ -42,26 +44,26 @@
+    public WriteResponseHandler(Replica replica, WriteType writeType, Runnable callback, long queryStartNanoTime)
     {
-        this(Arrays.asList(endpoint), Collections.<InetAddressAndPort>emptyList(), ConsistencyLevel.ONE, null, callback, writeType, queryStartNanoTime);
+        this(new ReplicaList(Collections.singleton(replica)), new ReplicaList(), ConsistencyLevel.ONE, null, callback, writeType, queryStartNanoTime);
```
Allocates an extra singleton, second replica list could be the immutable empty list instead of allocating.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/service/reads/AbstractReadExecutor.java:
```diff
@@ -269,24 +272,26 @@ public void executeAsync()
                 // not doing read repair; all replies are important, so it doesn't matter which nodes we
                 // perform data reads against vs digest.
-                makeDataRequests(initialReplicas.subList(0, 1));
+                makeDataRequests(initialReplicas.subList(0, 1).asEndpoints());
```
ImmutableList.of(intialReplicas.get(0).getEndpoint())

aweisberg commented on src/java/org/apache/cassandra/service/reads/AbstractReadExecutor.java:
```diff
@@ -269,24 +272,26 @@ public void executeAsync()
+                makeDataRequests(initialReplicas.subList(0, 1).asEndpoints());
                 if (initialReplicas.size() > 1)
-                    makeDigestRequests(initialReplicas.subList(1, initialReplicas.size()));
+                    makeDigestRequests(initialReplicas.subList(1, initialReplicas.size()).asEndpoints());
```
Avoid intermediate sublist

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/service/reads/AbstractReadExecutor.java:
```diff
@@ -344,17 +351,18 @@ public void maybeTryAdditionalReplicas()
         {
-            makeDataRequests(targetReplicas.subList(0, targetReplicas.size() > 1 ? 2 : 1));
+            Replicas.checkFull(targetReplicas);
+            makeDataRequests(targetReplicas.subList(0, targetReplicas.size() > 1 ? 2 : 1).asEndpoints());
```
Same thing with these two, can you avoid creating the extra intermediate list.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/service/reads/AbstractReadExecutor.java:
```diff
@@ -269,24 +272,26 @@ public void executeAsync()
                 // preferred by the snitch, we do an extra data read to start with against a replica more
                 // likely to reply; better to let RR fail than the entire query.
-                makeDataRequests(initialReplicas.subList(0, 2));
+                makeDataRequests(initialReplicas.subList(0, 2).asEndpoints());
```
Can you do a subList that is just the endpoint subList without converting it in a second step? Or should initialReplicas just be unwrapped immediately?

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/service/reads/DataResolver.java:
```diff
@@ -30,21 +36,23 @@
         this.enforceStrictLiveness = command.metadata().enforceStrictLiveness();
+
+        replicaMap = Maps.newHashMapWithExpectedSize(replicas.size());
+        replicas.forEach(r -> replicaMap.put(r.getEndpoint(), r));
```
Allocates a lambda (and an iterator).

bdeggleston replied:
fixed

aweisberg commented on test/long/org/apache/cassandra/locator/DynamicEndpointSnitchLongTest.java:
```diff
@@ -54,19 +54,19 @@ public void testConcurrency() throws InterruptedException, IOException, Configur
             for (int i = 0; i < 100; i++)
                 for (int j = 0; j < 256; j++)
-                    hosts.add(InetAddressAndPort.getByAddress(new byte[]{ 127, 0, (byte)i, (byte)j}));
+                    replicas.add(Replica.fullStandin(InetAddressAndPort.getByAddress(new byte[]{ 127, 0, (byte)i, (byte)j})));
```
Full standin, can we avoid it? Seems like the snitch is generally not concerned with either transientness or ranges just endpoints so we should need to have it using replicas?

bdeggleston replied:
fixed, the places that need it now deal with SystemReplicas.

aweisberg reviewed (changes requested):
We will definitely need unit tests for the new classes especially collections.

I bike shedded the daylights out of increases in allocations and copies. #1 focus is in collection methods which are the lowest level and used in the tightest loops when routing requests. #2 focus is allocations we perform one or more times to service either a read request or a write request.

The problem with escape analysis is that it tends to give up frequently as method size increases.

There is no need to optimize cases like starting compaction, repair, add/remove node etc since any additional overhead there occurs rarely.

I think I need to do another pass looking for bugs related to Replicas that were interacted with by address, but are now interacted with by the entire replica and that can result in contains and remove not working as expected. I don't think it happens often because in most cases the ranges don't vary but...

aweisberg commented:
Also I haven't reviewed the test changes yet.

aweisberg commented on src/java/org/apache/cassandra/locator/TokenMetadata.java:
```diff
@@ -904,38 +905,38 @@ private static PendingRangeMaps calculatePendingRanges(AbstractReplicationStrate
         {
             //Calculate all the ranges which will could be affected. This will include the ranges before and after the move.
-            Set<Range<Token>> moveAffectedRanges = new HashSet<>();
+            Set<Replica> moveAffectedReplicas = new HashSet<>();
```
Need to think carefully about this usage of Set<Replica> move is a case where endpoints and transientness changes but ranges might not. That's when it breaks down to try and use sets and set difference of all 3 attributes.

aweisberg commented on src/java/org/apache/cassandra/locator/PendingRangeMaps.java:
```diff
@@ -23,196 +23,176 @@
 import java.util.*;
 
-public class PendingRangeMaps implements Iterable<Map.Entry<Range<Token>, List<InetAddressAndPort>>>
+public class PendingRangeMaps implements Iterable<Map.Entry<Range<Token>, List<Replica>>>
```
Should this be consistent and stick to ReplicaList so we don't accidentally emit List<Replica> into the rest of Cassandra?

bdeggleston replied:
fixed

aweisberg commented on test/unit/org/apache/cassandra/locator/NetworkTopologyStrategyTest.java:
```diff
@@ -36,12 +37,17 @@
 import org.apache.cassandra.locator.TokenMetadata.Topology;
 import org.apache.cassandra.service.StorageService;
 
+import static org.apache.cassandra.locator.Replica.full;
```
Generally not a fan of static importing a single method. At least not this one.

bdeggleston replied:
fixed

bdeggleston committed:
doc review changes

bdeggleston committed:
using collection views instead of allocating a new collection

bdeggleston committed:
collection improvements

bdeggleston committed:
more collection improvements

bdeggleston committed:
adding is local to Replica

aweisberg commented:
Hey one thing I found is that the various implementations of ReplicaSet and ReplicaList (immutable, singleton) don't handle equals and hash code correctly.

You probably want an AbstractReplicaSet and AbstractReplicaList with no storage that define how equals (against any derived class of AbstractReplicaSet or List depending).

aweisberg commented:
Uh I may have brain farted there. I don't think there is an issue.

aweisberg commented:
I'm not totally crazy. I added a ReplicaSet.of method and when I call it I get Replicas.of instead which is a little nuts.

What I don't get is I am explicit invoking ReplicaSet and it's a static method! This is some of the craziest language behavior I have ever seen.

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaSet.java:
```diff
@@ -0,0 +1,156 @@
+    @Override
+    public void removeEndpoint(InetAddressAndPort endpoint)
+    {
+        for (Replica replica: replicaSet)
```
This is a CME because it keeps iterating even after it has found the thing to remove. It needs to keep iterating to be correct.

bdeggleston replied:
fixed

aweisberg commented on src/java/org/apache/cassandra/locator/ReplicaList.java:
```diff
@@ -0,0 +1,283 @@
+     * For allocating ReplicaLists where the final size is unknown, but
+     * should be less than the given size. Prevents overallocations in cases
+     * where there are less than the default ArrayList size, and defers to the
+     * ArrayList algorithem where there might be more
```
typo in algorithm

bdeggleston replied:
fixed

bdeggleston committed:
review fixes

bdeggleston committed:
review fixes

bdeggleston committed:
making PendingRangeMaps use ReplicaList

bdeggleston committed:
moving replicas equality up to base class

bdeggleston commented:
bq. You probably want an AbstractReplicaSet and AbstractReplicaList with no storage that define how equals (against any derived class of AbstractReplicaSet or List depending).

fixed, moved equals and hashCode up to base class

bdeggleston committed:
removing `fullStandIn` and the other static contstruction methods

bdeggleston committed:
splitting replica collection and convenience method classes

aweisberg commented:
Reminder we wanted to disallow the creation of transiently replicated keyspaces in mixed version clusters.

aweisberg commented on src/java/org/apache/cassandra/locator/Replicas.java:
```diff
@@ -50,6 +50,30 @@
     protected abstract Collection<Replica> getUnmodifiableCollection();
 
+
+    public boolean equals(Object o)
```
Sorry I should have been clearer. We would need AbstractReplicaList and AbstractReplicaSet to each implement their own equals and hash code. If you look at Java's AbstractList and AbstractSet you will see they implement their own equals and hashCode where the two aren't ever equal.

And I think it's not just cargo culting to also do it that way. I think it makes sense they should never be equal because sets have unpredictable iteration order and it leads to bugs to have it work only some of the time. Sets  also use a hash code specific to sets which is the sum of hash codes of the elements which makes it order insensitive.

aweisberg replied:
So I want to try and be clearer again :-P 

ReplicaList and ReplicaSet both work fine the way they are today. What doesn't work is Replicas functions that return an ImmutableContainer because those don't implement equals or hashCode for their contents. We should never return a Replicas that isn't either an instance of either ReplicaSet or ReplicaList with the implied equality and hashCode behavior. 

Java has the Collections class with singleton (which is singleton set) singletonList, singletonMap etc. Guava has ImmutableXYZ and the .of methods.

I think Guava got it right having those methods be in the type specific class rather than just a generic Collections class.

Otherwise it's all good. ReplicaSet and ReplicaList are wrappers for actual set and list implementations so it's not clear we would ever need to have more than just that. I think optimized singleton implementations should be a later optimization unless you feel driven.

bdeggleston replied:
Good point on set/list equality. Reverted.

>  We should never return a Replicas that isn't either an instance of either ReplicaSet or ReplicaList 

Why not? The methods that return the immutable containers are basically just returning transformed views of their arguments. The fact that they return generic `ReplicaCollection` instances is a pretty clear indicator that you can't expect them to behave like lists or sets.

aweisberg replied:
Because it breaks equality comparisons and hash code as they don't really have a good definition. If you look at how java.util.Collections and Guava do it you can't actually construct something that isn't either a set or a list. I think it's a better approach then having some things returned that aren't comparable to other collections (either list or set) with the same contents.

bdeggleston replied:
> If you look at how java.util.Collections and Guava do it you can't actually construct something that isn't either a set or a list.

The objects returned by com.google.common.collect.Iterables (which these methods are imitating) don't define equals or hashCode. 

The issue here might be that I'm thinking of these objects as iterables, and you seem to be thinking of them as  collections. Maybe this is more a naming issue?

aweisberg replied:
Sure if they don't implement they implement Iterable instead of Collection I think it's fine for them to not be an implementation of a collection (list, set, multi-set).

Interesting side note, java.util.AbstractQueue doesn't implement equals! Neither does say ArrayBlockingQueue or ArrayDeque.

aweisberg commented on src/java/org/apache/cassandra/locator/Replicas.java:
```diff
@@ -50,6 +50,30 @@
+
+    public int hashCode()
+    {
+        int result = 17;
```
Result starts as 17? I don't recall seeing that being done normally.

bdeggleston replied:
that's how it's done in Effective Java, bottom of page 48

aweisberg commented:
Can we get more thorough tests for Replicas, ReplicaList, and ReplicaSet. I'm not looking for 100% coverage, but things like equality and hash code, order sensitivity of hash code output and equals results.

ReplicaList, containsEndpoint, filter, intersect, subList, normalizeByRange (I am not sure this is equivalent to Range.normalize they look very different).

Then tests for methods in Replicas like containsEndpoint.

aweisberg commented:
Some test failures https://circleci.com/gh/aweisberg/cassandra/1223#tests/containers/60
https://circleci.com/gh/aweisberg/cassandra/1225#tests/containers/46
The dtest failure I suspect is it failing to notice the node move and push the notification. I haven't run it multiple times yet to see if it is flakey or not.

aweisberg commented:
Sankalp made the suggestion that we should forbid altering transient replication during range movements and other things that might conflict. Currently we allow RF changes without really checking if it is safe to do so and there is maybe no reason to continue that.

ifesdjeen commented on src/java/org/apache/cassandra/cql3/statements/AlterKeyspaceStatement.java:
```diff
@@ -96,7 +98,35 @@ private void warnIfIncreasingRF(KeyspaceMetadata ksm, KeyspaceParams params)
+
+    private void validateTransientReplication(AbstractReplicationStrategy oldStrategy, AbstractReplicationStrategy newStrategy)
+    {
+        if (oldStrategy.getReplicationFactor().trans == 0 && newStrategy.getReplicationFactor().trans > 0)
```
We can compare with `!=` (possibly only in the second case).

ifesdjeen commented on src/java/org/apache/cassandra/db/ConsistencyLevel.java:
```diff
@@ -148,40 +150,45 @@ public boolean isLocal(InetAddressAndPort endpoint)
+        return isLocal(replica.getEndpoint());
+    }
+
+    public int countLocalEndpoints(Iterable<Replica> liveReplicas)
```
This can be just `Replicas`

ifesdjeen commented on src/java/org/apache/cassandra/db/ConsistencyLevel.java:
```diff
@@ -148,40 +150,45 @@ public boolean isLocal(InetAddressAndPort endpoint)
     }
 
-    private Map<String, Integer> countPerDCEndpoints(Keyspace keyspace, Iterable<InetAddressAndPort> liveEndpoints)
+    private Map<String, Integer> countPerDCEndpoints(Keyspace keyspace, Iterable<Replica> liveReplicas)
```
This can be just `Replicas`

ifesdjeen commented on src/java/org/apache/cassandra/db/ConsistencyLevel.java:
```diff
@@ -190,50 +197,50 @@ public int countLocalEndpoints(Iterable<InetAddressAndPort> liveEndpoints)
     }
 
-    private List<InetAddressAndPort> filterForEachQuorum(Keyspace keyspace, List<InetAddressAndPort> liveEndpoints)
+    private ReplicaList filterForEachQuorum(Keyspace keyspace, ReplicaList liveReplicas)
```
Here, the sorted variant is not preserved (even though we return a sorted collection from `filterForQuery`).

We could retain sorted invariant if we used `filter`, maybe something like:

```
    private ReplicaList filterForEachQuorum(Keyspace keyspace, ReplicaList liveReplicas)
    {
        NetworkTopologyStrategy strategy = (NetworkTopologyStrategy) keyspace.getReplicationStrategy();
        Map<String, Integer> dcsReplicas = new HashMap<>();
        for (String dc : strategy.getDatacenters())
        {
            // we put _up to_ dc replicas only
            dcsReplicas.put(dc, localQuorumFor(keyspace, dc));
        }

        return liveReplicas.filter((replica) -> {
            String dc = DatabaseDescriptor.getEndpointSnitch().getDatacenter(replica);
            int replicas = dcsReplicas.get(dc);
            if (replicas > 0)
            {
                dcsReplicas.put(dc, --replicas);
                return true;
            }
            return false;
        });
    }
```

(also would avoid three iterations in favour of just two).

ifesdjeen commented on src/java/org/apache/cassandra/db/ConsistencyLevel.java:
```diff
@@ -190,50 +197,50 @@ public int countLocalEndpoints(Iterable<InetAddressAndPort> liveEndpoints)
     }
 
-    public boolean isSufficientLiveNodes(Keyspace keyspace, Iterable<InetAddressAndPort> liveEndpoints)
+    public boolean isSufficientLiveNodes(Keyspace keyspace, Iterable<Replica> liveReplicas)
```
We can use `Replicas` here

ifesdjeen commented on src/java/org/apache/cassandra/db/ConsistencyLevel.java:
```diff
@@ -242,11 +249,11 @@ public boolean isSufficientLiveNodes(Keyspace keyspace, Iterable<InetAddressAndP
     }
 
-    public void assureSufficientLiveNodes(Keyspace keyspace, Iterable<InetAddressAndPort> liveEndpoints) throws UnavailableException
+    public void assureSufficientLiveNodes(Keyspace keyspace, Iterable<Replica> liveReplicas) throws UnavailableException
```
We can use `Replicas` here.

Another question: should there be a distinction between sufficient live nodes for read and write path? Do we want to make sure there's a sufficient amount of live full nodes here or later?

ifesdjeen commented on src/java/org/apache/cassandra/dht/RangeStreamer.java:
```diff
@@ -259,36 +266,36 @@ private boolean useStrictSourcesForRanges(String keyspaceName)
             {
                 if (preEntry.getKey().contains(desiredRange))
                 {
-                    Set<InetAddressAndPort> oldEndpoints = Sets.newHashSet(preEntry.getValue());
```
Here, we're creating a new `Set`, then removing things from it. We could do an equivalent (and avoid copy of `newEndpoints`) by just running `filter` on `preEntry.getValue()`.

ifesdjeen commented on src/java/org/apache/cassandra/locator/AbstractReplicationStrategy.java:
```diff
@@ -102,35 +104,35 @@ protected AbstractReplicationStrategy(String keyspaceName, TokenMetadata tokenMe
-            endpoints = new ArrayList<InetAddressAndPort>(calculateNaturalEndpoints(searchToken, tm));
-            cachedEndpoints.put(keyToken, endpoints);
+            endpoints = calculateNaturalReplicas(searchToken, tm);
+            cachedReplicas.put(keyToken, endpoints);
```
It seems that we're only putting here, so here we can have an immutable collection already, which might spare us a need for copy in `getCachedReplicas` consumers.

ifesdjeen commented on src/java/org/apache/cassandra/locator/AbstractReplicationStrategy.java:
```diff
@@ -102,35 +104,35 @@ protected AbstractReplicationStrategy(String keyspaceName, TokenMetadata tokenMe
         }
 
-        return new ArrayList<InetAddressAndPort>(endpoints);
+        return new ReplicaList(endpoints);
```
E.g. this copy could be spared if we knew that cached endpoints contains an immutable collection.

ifesdjeen commented on src/java/org/apache/cassandra/locator/AbstractReplicationStrategy.java:
```diff
@@ -329,6 +335,10 @@ public static void validateReplicationStrategy(String keyspaceName,
         AbstractReplicationStrategy strategy = createInternal(keyspaceName, strategyClass, tokenMetadata, snitch, strategyOptions);
         strategy.validateExpectedOptions();
         strategy.validateOptions();
+        if (strategy.getReplicationFactor().trans > 0 && !DatabaseDescriptor.isTransientReplicationEnabled())
```
Check for version. Previously we had allNodesAtLeast22/30 for some internal handling, so we might have to implement similar mechanism for that.

ifesdjeen commented on src/java/org/apache/cassandra/locator/NetworkTopologyStrategy.java:
```diff
@@ -90,41 +98,53 @@ public NetworkTopologyStrategy(String keyspaceName, TokenMetadata tokenMetadata,
+        int transients;
 
-        DatacenterEndpoints(int rf, int rackCount, int nodeCount, Set<InetAddressAndPort> endpoints, Set<Pair<String, String>> racks)
+        DatacenterEndpoints(ReplicationFactor rf, int rackCount, int nodeCount, ReplicaSet replicas, Set<Pair<String, String>> racks)
```
Should we call it `DatacenterReplicas` now?

ifesdjeen commented on src/java/org/apache/cassandra/locator/NetworkTopologyStrategy.java:
```diff
@@ -90,41 +98,53 @@ public NetworkTopologyStrategy(String keyspaceName, TokenMetadata tokenMetadata,
             if (done())
                 return false;
 
+            if (replicas.containsEndpoint(ep))
```
Previously this has been handled with `endpoints.add` which was using Set as well, and it was throwing with `assert` rather than skipping. 

Did we change some subtle semantics here?

ifesdjeen commented on src/java/org/apache/cassandra/locator/OldNetworkTopologyStrategy.java:
```diff
@@ -36,27 +36,30 @@
         Token primaryToken = iter.next();
-        endpoints.add(metadata.getEndpoint(primaryToken));
+        Token previousToken = metadata.getPredecessor(primaryToken);
+        assert rf.trans == 0: "support transient replicas";
```
Should we move an assert to the top?

aweisberg replied:
Also if it's not supported should it be possible to disable the check?

ifesdjeen commented on src/java/org/apache/cassandra/locator/ReplicaList.java:
```diff
@@ -0,0 +1,270 @@
+        return new ReplicaList();
+    }
+
+    public static ReplicaList fullStandIns(Collection<InetAddressAndPort> endpoints)
```
`Collection` is already `Iterable`, so we only need a looser version since they seem to be identical.

ifesdjeen commented on src/java/org/apache/cassandra/locator/ReplicaList.java:
```diff
@@ -0,0 +1,270 @@
+     * where there are less than the default ArrayList size, and defers to the
+     * ArrayList algorithm where there might be more
+     */
+    public static ReplicaList withMaxSize(int size)
```
Naming seems to suggest that it can't grow over `size`. It's more `initialSize`, or?

ifesdjeen commented on src/java/org/apache/cassandra/locator/ReplicaMultimap.java:
```diff
@@ -0,0 +1,132 @@
+
+public abstract class ReplicaMultimap<K, V extends Replicas>
+{
+    Map<K, V> map = new HashMap<>();
```
I think we can make it final.

ifesdjeen commented on src/java/org/apache/cassandra/locator/ReplicationFactor.java:
```diff
@@ -0,0 +1,121 @@
+    {
+        Preconditions.checkArgument(trans == 0 || DatabaseDescriptor.isTransientReplicationEnabled(),
+                                    "Transient replication is not enabled on this node");
+        Preconditions.checkArgument(replicas >= 0,
```
To my understanding, replication factor has to be strictly positive (e.g. min of 1).

ifesdjeen commented on src/java/org/apache/cassandra/locator/ReplicationFactor.java:
```diff
@@ -0,0 +1,121 @@
+                                    "Transient replication is not enabled on this node");
+        Preconditions.checkArgument(replicas >= 0,
+                                    "Replication factor must be non-negative, found %s", replicas);
+        Preconditions.checkArgument(trans == 0 || trans < replicas,
```
We should also check for nonnegative here, since `full` is calculated by `replicas - trans`

ifesdjeen commented on src/java/org/apache/cassandra/locator/TokenMetadata.java:
```diff
@@ -1204,21 +1205,21 @@ private String printPendingRanges()
     }
 
-    public Collection<InetAddressAndPort> pendingEndpointsFor(Token token, String keyspaceName)
+    public Replicas pendingEndpointsFor(Token token, String keyspaceName)
```
If we return with more specific type, but pass less specific type, it might be simpler to establish a contract for List and Set.

ifesdjeen commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -78,7 +78,6 @@
 import org.apache.cassandra.service.paxos.ProposeVerbHandler;
 import org.apache.cassandra.net.MessagingService.Verb;
 import org.apache.cassandra.tracing.Tracing;
-import org.apache.cassandra.transport.Server;
```
`CacheLoder` `hintsInProgress` can be from `Replica` to `AtomicInteger`, this way we can get rid of many mentions of `InetAddressAndPort`

aweisberg replied:
Replica is also a range? Hints in progress isn't by range.

ifesdjeen commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -344,47 +343,43 @@ private static void recordCasContention(int contentions)
         Token tk = key.getToken();
-        List<InetAddressAndPort> naturalEndpoints = StorageService.instance.getNaturalEndpoints(metadata.keyspace, tk);
-        Collection<InetAddressAndPort> pendingEndpoints = StorageService.instance.getTokenMetadata().pendingEndpointsFor(tk, metadata.keyspace);
+        ReplicaList naturalReplicas = StorageService.instance.getNaturalReplicas(metadata.keyspace, tk);
```
Looks like we can just concat natural (as list) and pending (as set), without additional copying in-between.

ifesdjeen commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -344,47 +343,43 @@ private static void recordCasContention(int contentions)
-            naturalEndpoints = ImmutableList.copyOf(Iterables.filter(naturalEndpoints, isLocalDc));
-            pendingEndpoints = ImmutableList.copyOf(Iterables.filter(pendingEndpoints, isLocalDc));
+            Predicate<Replica> isLocalDc = sameDCPredicateFor(localDc);
+            naturalReplicas = ReplicaList.immutableCopyOf(naturalReplicas.filter(isLocalDc));
```
Do we need to make it immutable here, since we concat it later.

ifesdjeen commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -344,47 +343,43 @@ private static void recordCasContention(int contentions)
-            throw new UnavailableException(consistencyForPaxos, requiredParticipants, liveEndpoints.size());
+
+        Replicas concatenated = Replicas.concatNaturalAndPending(naturalReplicas, pendingReplicas);
+        ReplicaList liveReplicas = ReplicaList.immutableCopyOf(Replicas.filter(concatenated, IAsyncCallback.isReplicaAlive));
```
Same here (with Immutable).

ifesdjeen commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -503,14 +498,14 @@ private static void sendCommit(Commit commit, Iterable<InetAddressAndPort> repli
+        for (Replica replica: replicas)
         {
-            if (canDoLocalRequest(target))
+            if (canDoLocalRequest(replica.getEndpoint()))
```
`canDoLocalRequest` can receive Replica (or we can move it to replica). This will spare many `getEndpoint` calls where this method is called.

ifesdjeen commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -3788,7 +3789,10 @@ public void forceTerminateAllRepairSessions()
             throw new IllegalArgumentException("Unknown table '" + cf + "' in keyspace '" + keyspaceName + "'");
 
-        return getNaturalEndpoints(keyspaceName, tokenMetadata.partitioner.getToken(metadata.partitionKeyType.fromString(key))).stream().map(i -> i.address).collect(toList());
+        ReplicaList replicas = getNaturalReplicas(keyspaceName, tokenMetadata.partitioner.getToken(metadata.partitionKeyType.fromString(key)));
```
We could use `replicas.asEndpointsList`, and return list of `InetAddressAndPort.

ifesdjeen commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -4231,53 +4211,53 @@ private void calculateToFromStreams(Collection<Token> newTokens, List<String> ke
                      * nodes in the ring responsible for data we are interested in
                      */
-                    Multimap<Range<Token>, InetAddressAndPort> rangesToFetchWithPreferredEndpoints = ArrayListMultimap.create();
+                    ReplicaMultimap<Range<Token>, ReplicaList> rangesToFetchWithPreferredEndpoints = ReplicaMultimap.list();
```
We can iterate over `entrySet` rather than call `get` below.

ifesdjeen commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -4231,53 +4211,53 @@ private void calculateToFromStreams(Collection<Token> newTokens, List<String> ke
                                 {
-                                    Set<InetAddressAndPort> oldEndpoints = Sets.newHashSet(rangeAddresses.get(range));
-                                    Set<InetAddressAndPort> newEndpoints = Sets.newHashSet(strategy.calculateNaturalEndpoints(toFetch.right, tokenMetaCloneAllSettled));
+                                    ReplicaSet oldEndpoints = new ReplicaSet(rangeAddresses.get(range));
```
Might be better to filter on `oldEndpoints` rather than remove from them (esp. since we already copy)

Additional benefit would be if we ever decide to switch to immutable interface, we'll have to do this anyways.

ifesdjeen commented on src/java/org/apache/cassandra/service/reads/DataResolver.java:
```diff
@@ -64,12 +75,19 @@ public PartitionIterator resolve()
         int count = responses.size();
         List<UnfilteredPartitionIterator> iters = new ArrayList<>(count);
-        InetAddressAndPort[] sources = new InetAddressAndPort[count];
+        Replica[] sources = new Replica[count];
```
This is the only usage of array (e.g. this place and dependent ones), and it seems it can be avoided in favour of list.

ifesdjeen commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -1275,36 +1272,38 @@ private static WriteResponseHandlerWrapper wrapViewBatchResponseHandler(Mutation
-                                             String localDataCenter,
-                                             Stage stage)
+    public static void sendToHintedReplicas(final Mutation mutation,
+                                            Iterable<Replica> targets,
```
We can use `Replicas` here

ifesdjeen commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -1275,36 +1272,38 @@ private static WriteResponseHandlerWrapper wrapViewBatchResponseHandler(Mutation
 
         // this dc replicas:
-        Collection<InetAddressAndPort> localDc = null;
+        Collection<Replica> localDc = null;
```
We can use `Replicas` here

ifesdjeen commented on src/java/org/apache/cassandra/db/view/ViewUtils.java:
```diff
@@ -58,46 +57,55 @@ private ViewUtils()
         // number of replicas for all of the tokens in the ring.
-        assert baseEndpoints.size() == viewEndpoints.size() : "Replication strategy should have the same number of endpoints for the base and the view";
-        int baseIdx = baseEndpoints.indexOf(FBUtilities.getBroadcastAddressAndPort());
+        assert baseReplicas.size() == viewReplicas.size() : "Replication strategy should have the same number of endpoints for the base and the view";
```
We could simplify this slightly by using set + iterator instead of `get` here: 

```diff
         String localDataCenter = DatabaseDescriptor.getEndpointSnitch().getDatacenter(FBUtilities.getBroadcastAddressAndPort());
-        ReplicaList baseReplicas = new ReplicaList();
-        ReplicaList viewReplicas = new ReplicaList();
+        ReplicaSet baseReplicas = new ReplicaSet();
+        ReplicaSet viewReplicas = new ReplicaSet();
+        // We might add a method that filters natural endpoints by dc
         for (Replica baseEndpoint : replicationStrategy.getNaturalReplicas(baseToken))
         {
             // An endpoint is local if we're not using Net
@@ -92,20 +95,18 @@ public final class ViewUtils
         // number of replicas for all of the tokens in the ring.
         assert baseReplicas.size() == viewReplicas.size() : "Replication strategy should have the same number of endpoints for the base and the view";

-        int baseIdx = -1;
-        for (int i=0; i<baseReplicas.size(); i++)
+        // we don't need "get" here, we can just use
+        Iterator<Replica> baseReplicaIteraror = baseReplicas.iterator();
+        Iterator<Replica> viewReplicaIteraror = viewReplicas.iterator();
+        while (baseReplicaIteraror.hasNext() && viewReplicaIteraror.hasNext())
         {
-            if (baseReplicas.get(i).isLocal())
-            {
-                baseIdx = i;
-                break;
-            }
+            Replica baseReplica = baseReplicaIteraror.next();
+            Replica viewReplica = viewReplicaIteraror.next();
+            if (baseReplica.isLocal())
+                return Optional.of(viewReplica);
         }

-        if (baseIdx < 0)
-            //This node is not a base replica of this key, so we return empty
-            return Optional.empty();
-
-        return Optional.of(viewReplicas.get(baseIdx));
+        //This node is not a base replica of this key, so we return empty
+        return Optional.empty();
     }
 }
```

One more reason not to use `Set` here is since there's `contains` that would use iteration internally.

ifesdjeen commented on src/java/org/apache/cassandra/locator/AbstractReplicationStrategy.java:
```diff
@@ -202,61 +204,65 @@ private Keyspace getKeyspace()
      * so we could cache/invalidate cleanly.)
      */
-    public Multimap<InetAddressAndPort, Range<Token>> getAddressRanges(TokenMetadata metadata)
+    public ReplicaMultimap<InetAddressAndPort, ReplicaSet> getAddressReplicas(TokenMetadata metadata)
```
Because of how it's used, we can add another, non-map variant, since in the majority of usages we're just calling `get` on just-materialised map.

ifesdjeen commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -4231,53 +4211,53 @@ private void calculateToFromStreams(Collection<Token> newTokens, List<String> ke
                 {
                     // getting collection of the currently used ranges by this keyspace
-                    Collection<Range<Token>> currentRanges = endpointToRanges.get(localAddress);
+                    ReplicaSet currentReplicas = endpointToRanges.get(localAddress);
```
`currentReplicas` is used inside the loop only.

ifesdjeen commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -4231,53 +4211,53 @@ private void calculateToFromStreams(Collection<Token> newTokens, List<String> ke
                 // replication strategy of the current keyspace
                 AbstractReplicationStrategy strategy = Keyspace.open(keyspace).getReplicationStrategy();
-                Multimap<InetAddressAndPort, Range<Token>> endpointToRanges = strategy.getAddressRanges();
+                ReplicaMultimap<InetAddressAndPort, ReplicaSet> endpointToRanges = strategy.getAddressReplicas();
```
add `getAddressReplicas` by `InetAddressAndPort` to avoid materialising map.

ifesdjeen commented on src/java/org/apache/cassandra/locator/TokenMetadata.java:
```diff
@@ -733,19 +733,19 @@ public InetAddressAndPort getEndpoint(Token token)
+                for (Replica replica : entry.getValue())
                 {
-                    map.put(range, address);
+                    map.put(range, replica);
```
Since we're putting by `range` and iterating over `entry.getValue` here, it seems like we can spare this iteration.

ifesdjeen commented on src/java/org/apache/cassandra/locator/ReplicaList.java:
```diff
@@ -0,0 +1,270 @@
+            ReplicaList normalized = new ReplicaList(size());
+            for (Replica replica: this)
+            {
+                replica.addNormalizeByRange(normalized);
```
I might misunderstand something, but woudln't it be simpler to read if we did `replica.normalize()` that would return a new `Replica`, which we would in turn add to `normalized` ourselves here?..

ifesdjeen commented on test/microbench/org/apache/cassandra/test/microbench/PendingRangesBench.java:
```diff
@@ -97,13 +103,13 @@ public void searchTokenForOldPendingRanges(final Blackhole bh)
-        Set<InetAddressAndPort> endpoints = new HashSet<>();
-        for (Map.Entry<Range<Token>, Collection<InetAddressAndPort>> entry : oldPendingRanges.asMap().entrySet())
+        Set<Replica> replicas = new HashSet<>();
+        for (Map.Entry<Range<Token>, Collection<Replica>> entry : oldPendingRanges.asMap().entrySet())
```
Since all usages of `asMap` are used with `entrySet`, should we just expose `entrySet` instead and keep map private?

ifesdjeen commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -1325,10 +1324,10 @@ public static void sendToHintedEndpoints(final Mutation mutation,
                     else
                     {
-                        Collection<InetAddressAndPort> messages = (dcGroups != null) ? dcGroups.get(dc) : null;
+                        Replicas messages = (dcGroups != null) ? dcGroups.get(dc) : null;
```
We can use `computeIfAbsent` here

ifesdjeen reviewed (changes requested):
Thank you for the patch! Great work!

Wanted to make several suggestions and wanted to make another (more correctness-oriented) pass after the comments are addressed, but this is a first round of reviews.

Some things that were not in the comments attached to code: 

  * It was mentioned that we have to forbid RF changes for now until this is tested (in comments), just keeping track of it
  * I would add tests for 2i / MV creation (and change to transient replication in 2i/MV clusters)
  * Generally, ConsistencyLevel is not tested directly, we could fix that (e.g. quorumfor/localquorumfor). It wasn't before, either, but we can use time while refactoring to improve this.
  * It might be flushed out in the future, but for now it’s unclear how “transient” part of RF is going to interact with the rest of the system. Do you think we'll ever need a separation between full and transient replicas, or will set/list and iteration over them always be enough?
  * `ReplicaSet` can be a map from endpoint to replica, which might help with both iteration and access to endpoints 
  * Generally, patch does not follow guidelines for if/else statements (single-line ifs aren’t wrapped with curly brackets), for example [https://github.com/apache/cassandra/pull/224/files#diff-e3a60bbc39d350c19b99e08d9d9af1efR113] and [https://github.com/apache/cassandra/pull/224/files#diff-683dc6a776598ace3a7981c8ca28670eR253] and [https://github.com/apache/cassandra/pull/224/files#diff-f489d24d98cac9147ea0b42a5a7f53a2R110]. All signs should be surrounded by whitespace (e.g. `a <= b` vs `a<=b`)

It'd also be great to hear your opinion on making clearer distinction between sorted and unsorted replica set/list and making replica list and set immutable (since this seems to be the way they're used predominantly). This can make code safer to and potentially simpler to extend.

bdeggleston committed:
Revert "moving replicas equality up to base class"

This reverts commit e7fac2c

aweisberg commented:
> ReplicaSet can be a map from endpoint to replica, which might help with both iteration and access to endpoints

Should it be keyed by endpoint or range? I don't think what we want is consistent.

aweisberg commented:
Additional note, forbid the use of counters with transient replication?
